Parse NOT NULL for MySQL generated columns

GeorchW opened the pull request:
### Description

As per the MySQL CREATE TABLE specification (see https://dev.mysql.com/doc/refman/8.0/en/create-table.html), the location of the NULL/NOT NULL syntax is slightly different for generated vs. regular columns: in generated columns, the NULL/NOT NULL goes after the VIRTUAL/STORED instead of directly after the data type.

Fixes #10850

### Checklist
<!-- go over following points. check them with an `x` if they do apply, (they turn into clickable checkboxes once the PR is submitted, so no need to do everything at once)

-->

This pull request is:

- [ ] A documentation / typographical / small typing error fix
	- Good to go, no issue or tests are needed
- [x] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [ ] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

GeorchW committed:
Parse NOT NULL for MySQL generated columns

As per the MySQL CREATE TABLE specification (see https://dev.mysql.com/doc/refman/8.0/en/create-table.html),
the location of the NULL/NOT NULL syntax is slightly different for generated vs. regular columns: in generated
columns, the NULL/NOT NULL goes after the VIRTUAL/STORED instead of directly after the data type.

Fixes #10850

GeorchW commented:
btw, the other dialects use the database's metadata views (which would be `information_schema` in MySQL) instead of parsing `CREATE TABLE` statements. Is this a compatibility thing in the MySQL dialect or something? Or is it just "not done yet"? It feels a little spooky to me to parse context-free languages with regular expressions, although I must acknowledge that it's probably "good enough" for the vast majority of cases (i.e. as long as expressions don't contain keywords in strings).

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 5cce2f4208e8b4cd179e67d668720feaa3ff5f04 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
New Gerrit review created for change 5cce2f4208e8b4cd179e67d668720feaa3ff5f04: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5102

zzzeek commented:
this PR is great!  thanks for just jumping on it!

GeorchW commented:
Force-pushed to (hopefully) make CI work, was just a missing trailing comma.

@zzzeek That makes sense. Didn't know about the situation in earlier MySQL.

GeorchW commented:
@CaselIT 
> Regarding the change, does mariadb behave the same?

[The MariaDB docs](https://mariadb.com/kb/en/generated-columns/) claim that the syntax is compatible, but they don't say anything about nullable generated columns as far as I can see. I just tried both variants:

```sql
CREATE TABLE IF NOT EXISTS mytable ( 
    my_generated_column int  GENERATED ALWAYS AS (1234) VIRTUAL NOT NULL 
);
CREATE TABLE IF NOT EXISTS mytable (
    my_generated_column int NOT NULL GENERATED ALWAYS AS (1234) VIRTUAL
);
```
Neither works. Tested on MariaDB 11.2.2 (via Docker). When leaving out the `NOT NULL` alltogether, the column is considered to be nullable.

So... yeah, that's another bug, I think it's not possible to create non-nullable generated columns with MariaDB+SQLAlchemy right now (not sure though). This PR shouldn't change anything about that, though.

sqla-tester reviewed (commented):
Hi, this is **sqla-tester** and I see you've pinged me for review. However, user **GeorchW** is not authorized to initiate CI jobs.  Please wait for a project member to do this!

CaselIT commented:
> So... yeah, that's another bug, I think it's not possible to create non-nullable generated columns with MariaDB+SQLAlchemy right now (not sure though).

Yes I remember something along there lines for one db.
Maybe mysql added support for it after sqlalchemy implemented computed?

In any case as long as there are no regression with the new regexp (and there should not be) on mariadb then it's fine

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 7cd93171299cc072fcdec581a422c0b9a7ab8d67 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset 7cd93171299cc072fcdec581a422c0b9a7ab8d67 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5102

sqla-tester commented:
**Michael Bayer** (zzzeek) wrote:

recheck

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5102

GeorchW commented:
https://jenkins.sqlalchemy.org/job/sqlalchemy_21_gerrit/cext=cext,db=mysql,pyv=py311/283/testReport/

Hmm, the tests complain that the specific syntax used in the test is not supported in MariaDB and old MySQL versions...

What's your opinion on how to deal with this? Can/should I somehow differentiate between the backends in the test, only adding the new test case for MySQL versions that actually support non-nullable generated columns?

GeorchW committed:
Fix tests (hopefully)

GeorchW commented:
I've adjusted the tests to check for the MySQL version, it seems like there are already similar checks in similar places. I think you can recheck if the approach is fine

CaselIT commented:
> https://jenkins.sqlalchemy.org/job/sqlalchemy_21_gerrit/cext=cext,db=mysql,pyv=py311/283/testReport/
> 
> Hmm, the tests complain that the specific syntax used in the test is not supported in MariaDB and old MySQL versions...
> 
> What's your opinion on how to deal with this? Can/should I somehow differentiate between the backends in the test, only adding the new test case for MySQL versions that actually support non-nullable generated columns?

looking at the update and at the fail I think this should be only a test constraint that we add, since the dialect has no use of that flag.

I'll suggest something later today so that the test automatically skips some test cases depending on the db in use

CaselIT commented:
refactoring that test does not seem to be warranted. Can you instead of creating an attribute use a property and maybe underscore it since it's a test only thing?

thanks!

GeorchW commented:
I made it an attribute since the other fields were attributes, too, but sure, I can change that

CaselIT commented:
well I was looking at the property `_is_mariadb_102` that's a similar one that used only by the test

GeorchW committed:
Convert _supports_notnull_generated_columns into property

GeorchW commented:
@CaselIT Ah right, gotcha. Adjusted it

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 0d58aaa81815264c571ca9ccaca3d256c48d74c7 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset 0d58aaa81815264c571ca9ccaca3d256c48d74c7 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5102

GeorchW committed:
Fix and refactor test_nullable_reflection

GeorchW commented:
Sorry, my bad, the provided code was broken; I had adjusted the test case to only be included for compatible MySQL versions, but did not alter the expected outcome accordingly.

I've now refactored the test a bit such that SQL DDL input and expected reflected properties are right next to each other, which IMO makes it a bit easier to read (but also a bit longer). Tell me what you think.

The tests take forever to execute on my machine. Even just the initialization of `tox` takes almost a minute. Is there a way to speed this up?

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision fb9a81020c393231ca90a1e88342b11cf64414a1 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset fb9a81020c393231ca90a1e88342b11cf64414a1 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5102

sqla-tester commented on test/dialect/mysql/test_reflection.py:
```diff
@@ -778,89 +778,140 @@ def test_nullable_reflection(self, metadata, connection):
+        else:
+            current_timestamp = "CURRENT_TIMESTAMP"
+
+        test_cases = [
```
**Michael Bayer** (zzzeek) wrote:

so if we were going to refactor this test, we should do it using `@testing.combinations`.

Here's the options for how we can proceed on this part.


1. we merge it as is, then I will go in after the fact and reorganize this into multiple, independent tests using `@testing.combinations`
2. the submitter Georg reorganizes this using `@testing.combinations` directly.  This is some new API to learn so I dont want to impose this on you unless you were super interested.
3.  I take over the gerrit here (we're not very good at updating github PRs) and I finish this up inline.   since I can work extremely quickly

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5102

GeorchW replied:
I'm not particularly keen on reorganizing the test case, so not having to do (2) would be nice. Otherwise, I don't really mind whether it's merged before or after

CaselIT replied:
no worries

sqla-tester replied:
**Michael Bayer** (zzzeek) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5102

sqla-tester reviewed (commented):
**Michael Bayer** (zzzeek) wrote:

so this needs a changelog file and I also want to refactor that test to use a more modern form which is `@testing.combinations`, this is a facade on top of pytest parametrize (https://docs.pytest.org/en/7.1.x/example/parametrize.html) that generates individual tests per case.    See inline comment for options on how to proceed.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5102

sqla-tester commented:
**Federico Caselli** (CaselIT) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5102

sqla-tester commented:
**Michael Bayer** (zzzeek) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5102

sqla-tester commented:
**Michael Bayer** (zzzeek) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5102

sqla-tester commented:
**Michael Bayer** (zzzeek) wrote:

that's an unrelated failure

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5121

sqla-tester commented:
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5121 has been **merged**. Congratulations! :)

GeorchW commented:
nice, thank you so much! I guess the next release that contains this change is going to happen in the next month or so?

CaselIT commented:
Yes, it's usually about once a month, so more or less in a couple of weeks
